frontend/lib/resolution/maybe-const.cpp

```diff
@@ -41,26 +41,56 @@ using namespace types;
 struct AdjustMaybeRefs {
   using RV = MutatingResolvedVisitor<AdjustMaybeRefs>;
 
-  typedef enum {
-    REF = 1,
-    CONST_REF = 2,
-    VALUE = 3,
-    REF_MAYBE_CONST = 4, // used temporarily for recursive cases
-  } Access;
+  // An abstracted view over a QualifiedType. Here 'OUT' and 'INOUT' kinds
+  // are viewed as being a form of 'ref'. Kinds that are not references are
+  // considered "VALUE" and generally ignored.
+  class Access {
+    QualifiedType type_;
+  public:
+    Access() = default;
+    Access(QualifiedType type) : type_(std::move(type)) {}
+    Access(QualifiedType::Kind q, const Type* t) : type_(QualifiedType(q, t)) {}
+    const QualifiedType& type() const { return type_; }
+    QualifiedType::Kind kind() const { return type_.kind(); }
+    bool isRef() const {
+      return kind() == QualifiedType::REF ||
+             kind() == QualifiedType::OUT ||
+             kind() == QualifiedType::INOUT;
+    }
+    bool isConstRef() const {
+      return kind() == QualifiedType::CONST_REF;
+    }
+    bool isRefMaybeConst() const {
+      return kind() == QualifiedType::REF_MAYBE_CONST;
```

Comment:
IIRC @jabraham17 has removed ref-maybe-const formals from the production compiler. I don't recall if they have gone through a whole deprecation cycle. But we should probably remove this logic from dyno. Could you make an issue about it?

Reply:
https://github.com/Cray/chapel-private/issues/6147